sky/spot/recovery_strategy.py

```diff
@@ -211,37 +196,47 @@ def recover(self) -> float:
             logger.info('Ignoring the job cancellation failure; the spot '
                         'cluster is likely completely stopped. Recovering.')
 
-        # Add region constraint to the task, to retry on the same region
-        # first.
-        task = self.dag.tasks[0]
-        resources = list(task.resources)[0]
-        original_resources = resources
-
-        launched_cloud = handle.launched_resources.cloud
-        launched_region = handle.launched_resources.region
-        new_resources = resources.copy(cloud=launched_cloud,
-                                       region=launched_region)
-        task.set_resources({new_resources})
-        # Not using self.launch to avoid the retry until up logic.
-        launched_time = self._launch(raise_on_failure=False)
-        # Restore the original dag, i.e. reset the region constraint.
-        task.set_resources({original_resources})
-        if launched_time is not None:
-            return launched_time
+        while True:
+            # Add region constraint to the task, to retry on the same region
+            # first.
+            task = self.dag.tasks[0]
+            resources = list(task.resources)[0]
+            original_resources = resources
+
+            launched_cloud = handle.launched_resources.cloud
+            launched_region = handle.launched_resources.region
+            new_resources = resources.copy(cloud=launched_cloud,
+                                        region=launched_region)
+            task.set_resources({new_resources})
+            # Not using self.launch to avoid the retry until up logic.
+            launched_time = self._launch(raise_on_failure=False)
+            # Restore the original dag, i.e. reset the region constraint.
+            task.set_resources({original_resources})
+            if launched_time is not None:
+                return launched_time
+
+            # Step 2
+            logger.debug('Terminating unhealthy spot cluster.')
+            self.terminate_cluster()
+
+            # Step 3
+            logger.debug('Relaunch the cluster  without constraining to prior '
+                        'cloud/region.')
+            # Not using self.launch to avoid the retry until up logic.
+            launched_time = self._launch(max_retry=self._MAX_RETRY_CNT,
+                                        raise_on_failure=False)
+            if launched_time is None:
+                # Failed to launch the cluster.
+                if self.retry_until_up:
+                    gap_seconds = self.RETRY_INIT_GAP_SECONDS
+                    logger.info(
+                        f'Retrying to recover the spot cluster in {gap_seconds:.1f} '
+                        'seconds.')
+                    time.sleep(gap_seconds)
+                    continue
```

Comment:
Why do we repeat Step 1 and Step 2 in this case? Worth a comment. (If steps 1 and 2 are not needed, presumably we can do `self._launch(max_retry=None if retry_until_up else self._MAX_RETRY_CNT, ...)` on L226.)

Reply:
Please refer to the comment above for the reason. : )